### PR TITLE
server: add server owned http route manager

### PR DESCRIPTION
**Pull request by junr03**
This PR adds a server owned HttpRouteManager, which observes the server's RDSRouteConfigProviders. This allows for deduplication of RDSRouteConfigProviders of the same name via shared ownership, and allows the server to iterate over all RDSRouteConfigProviders.

### CODE REVIEW

File: source/server/http_route_manager_impl.cc [hunk 1]
```diff
@@ -0,0 +1,34 @@
+#include "server/http_route_manager_impl.h"
+
+namespace Envoy {
+namespace Server {
+
+std::vector<Router::RouteConfigProviderSharedPtr> HttpRouteManagerImpl::routeConfigProviders() {
+  std::vector<Router::RouteConfigProviderSharedPtr> ret;
+  ret.reserve(route_config_providers_.size());
+  for (const auto& element : route_config_providers_) {
+    ret.push_back(element.second.lock());
```
Comment (junr03) [inline source/server/http_route_manager_impl.cc]: @mattklein123 is it possible here that between the time that the last shared_ptr for the provider gets reset, and the destructor for the provider is called (thus cleaning the entry in the manager), this iteration happens?

In other words, should I lock, check for nullptr, and then push into the vector only is the shared_ptr is not null?

Comment (mattklein123) [inline source/server/http_route_manager_impl.cc]: The whole implementation will need to change a bit so let's touch on this later.

File: include/envoy/server/filter_config.h [hunk 2]
```diff
@@ -58,6 +59,11 @@ class FactoryContext {
   virtual bool healthCheckFailed() PURE;
 
   /**
+   * @return HttpRouteManager& singleton for use by the entire server.
+   */
+  virtual HttpRouteManager& httpRouteManager() PURE;
```
Comment (mattklein123) [inline include/envoy/server/filter_config.h]: At the factory context level, we should not expose the entire manager. Only the ability to get a route provider by name. See below for interface changes.

File: include/envoy/server/http_route_manager.h [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+  /**
+   * Add a RouteConfigProvider. Ownership of the RouteConfigProvider is shared by all the
+   * ConnectionManagers who own a RouteConfigProviderSharedPtr. The HttpRouteManager holds weak_ptrs
+   * to the RouteConfigProviders.
+   * @param name supplies the name of the RouteConfigProvider. This is a unique identifier.
+   * @param routeConfigProvider is a weak_ptr to the RouteConfigProvider to add.
+   */
+  virtual void
+  addRouteConfigProvider(const std::string& name,
```
Comment (mattklein123) [inline include/envoy/server/http_route_manager.h]: add/get/remove should be merged into a single API so that RAII can be used when the returned provider is destroyed. Basically, I would recommend doing the following:

- Return a shared_ptr to a route provider (or interface handle to the route provider).
- Internally inside the manager, store *both* a weak_ptr and a raw pointer. The raw pointer can be used to lookup the manager entry to remove when the last shared_ptr reference goes away. The weak_ptr can be used to create a new shared_ptr to return if needed (AFAIK there is no way to get the raw pointer out of a weak_ptr making this necessary).
- This will allow you to have a single get() method exposed.

File: include/envoy/server/filter_config.h (hunk 2 above)
Comment (junr03) [inline include/envoy/server/filter_config.h]: @mattklein123 I have changed the HttpRouteManager interface like we talked about to only expose `get` and `list` the configured route providers. 

What is the pattern to follow here to not expose the entire manager? is it to have a `getRouteConfigProvider` method in this interface, or now that we can only get and list, is it fine to expose the manager at the context level?

Comment (mattklein123) [inline include/envoy/server/filter_config.h]: there is no reason to expose list() at the context level. Just have two interfaces, with one deriving from the other, and expose the first from the context.

Comment (junr03) [inline include/envoy/server/filter_config.h]: makes sense

Comment (junr03) [inline include/envoy/server/filter_config.h]: updated. I am not thrilled by the naming here. I'll think a bit more.

File: source/common/router/rds_impl.h [hunk 4]
```diff
@@ -49,75 +49,5 @@ class StaticRouteConfigProviderImpl : public RouteConfigProvider {
   ConfigConstSharedPtr config_;
 };
 
-/**
```
Comment (mattklein123) [inline source/common/router/rds_impl.h]: @junr03 to make reviewing this easier, can you please split out the file move into a separate PR with no code changes. Then do the PR with the changes? It's really hard to tell what changed. Or just don't move things in this PR and clean it up in another PR. Thanks.

**Comment by junr03** [comment]
@mattklein123 updated by making the manager part of the router namespace. Right now the manager is in the rds_impl file because I was getting tangled in the imports and forward declarations necessary to make it happen in an independent http_route_manager file.

File: include/envoy/router/http_route_manager.h [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+   * @param dispatcher supplies the dispatcher.
+   * @param random supplies the random generator.
+   * @param local_info supplies the local info.
+   * @param scope supplies the scope to use for the route config provider.
+   * @param stat_prefix supplies the stat_prefix to use for the provider stats.
+   * @param tls supplies the tls slot allocator.
+   * @param init_manager supplies the init manager.
+   */
+  virtual RouteConfigProviderSharedPtr
+  getRouteConfigProvider(const Json::Object& config, Runtime::Loader& runtime,
```
Comment (mattklein123) [inline include/envoy/router/http_route_manager.h]: Quick comment to work on (sorry really busy this week so I'm giving you drive by reviews :( ): From an interface perspective, I think almost all of these things don't need to be passed. They can be known internally to the route manager. I wouldn't pass any json either, and just unpack into exactly what you need which I think is just route table name and cluster name. (This will likely get converted to v2 but the conversion will be easy cc @htuch ). 

So I think at quick glance you need to pass cluster name, route table name, init manager, scope, and maybe stat_prefix.

Comment (junr03) [inline include/envoy/router/http_route_manager.h]: No worries, I should have seen that myself. After I update I will ping @htuch to see if he can give a review to load balance you a bit.

Comment (mattklein123) [inline include/envoy/router/http_route_manager.h]: This is close enough. Please just add tests and get the PR ready to ship with full coverage and then we can do a full review.

**Comment by junr03** [comment]
@mattklein123 @htuch tests are in. This is ready for review

File: include/envoy/router/http_route_manager.h [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+#include "envoy/stats/stats.h"
+#include "envoy/thread_local/thread_local.h"
+#include "envoy/upstream/cluster_manager.h"
+
+namespace Envoy {
+namespace Router {
+
+/**
+ * The HttpRouteManager exposes the ability to get a RouteConfigProvider. This interface is exposed
+ * to the Server's FactoryContext in order to allow HttpConnMans to get RouteConfigProviders.
```
Comment (htuch) [inline include/envoy/router/http_route_manager.h]: Nit: `HttpConnectionManagers`, prefer not to abbreviate type names.

File: include/envoy/router/http_route_manager.h [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+#include "envoy/runtime/runtime.h"
+#include "envoy/stats/stats.h"
+#include "envoy/thread_local/thread_local.h"
+#include "envoy/upstream/cluster_manager.h"
+
+namespace Envoy {
+namespace Router {
+
+/**
+ * The HttpRouteManager exposes the ability to get a RouteConfigProvider. This interface is exposed
```
Comment (htuch) [inline include/envoy/router/http_route_manager.h]: Nit: Prefer the more verbose `RouteConfigProviderManager`.

File: include/envoy/router/http_route_manager.h [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+ * to the Server's FactoryContext in order to allow HttpConnMans to get RouteConfigProviders.
+ */
+class HttpRouteManager {
+public:
+  virtual ~HttpRouteManager() {}
+
+  /**
+   * get a Router::RouteConfigProviderSharedPtr. Ownership of the RouteConfigProvider is shared by
```
Comment (htuch) [inline include/envoy/router/http_route_manager.h]: Nit: s/get/Get.

File: include/envoy/router/http_route_manager.h [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+   * all the ConnectionManagers who own a Router::RouteConfigProviderSharedPtr. The HttpRouteManager
+   * holds weak_ptrs to the RouteConfigProviders. Clean up of the weak ptrs happen from the
+   * destructor of the RouteConfigProvider. This function creates a RouteConfigProvider is there
```
Comment (htuch) [inline include/envoy/router/http_route_manager.h]: Nit: s/is/if

File: include/envoy/router/http_route_manager.h [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+ */
+class ServerHttpRouteManager : public HttpRouteManager {
+public:
+  virtual ~ServerHttpRouteManager() {}
+
+  /**
+   * @return std::vector<Router::RouteConfigProviderSharedPtr> a list of all the
+   * RouteConfigProviders currently loaded.
+   */
+  virtual std::vector<RouteConfigProviderSharedPtr> routeConfigProviders() PURE;
```
Comment (htuch) [inline include/envoy/router/http_route_manager.h]: Is this only for test? Maybe add a `ForTest` suffix. Do we need to return a proper shared ptr here, or is a reference good enough?

File: source/common/router/rds_impl.cc [hunk 11]
```diff
@@ -119,5 +125,45 @@ void RdsRouteConfigProviderImpl::registerInitTarget(Init::Manager& init_manager)
...
+  return ret;
+};
+
+Router::RouteConfigProviderSharedPtr HttpRouteManagerImpl::getRouteConfigProvider(
+    const Json::Object& config, Upstream::ClusterManager& cm, Stats::Scope& scope,
+    const std::string& stat_prefix, Init::Manager& init_manager) {
+
+  // RdsRouteConfigProviders are unique based on their <route_config_name>_<cluster>.
+  std::string map_identifier = config.getString("route_config_name") + "_";
+  map_identifier += config.getString("cluster");
```
Comment (htuch) [inline source/common/router/rds_impl.cc]: 1. There's some duplicated logic in creating the map key with the erase in the destructor, probably best to refactor.
2. Is it possible that as a result of `_` appearing in either cluster or route config name, we could get unintended clashes here? e.g. cluster "A_B" and route config name "C" would hit the same slot as cluster "A" and route config name "B_C".
3. Nit: prefer creating `map_identifier` in one expression and making it `const.

File: source/common/router/rds_impl.cc [hunk 12]
```diff
@@ -119,5 +125,45 @@ void RdsRouteConfigProviderImpl::registerInitTarget(Init::Manager& init_manager)
...
+                                           const LocalInfo::LocalInfo& local_info,
+                                           ThreadLocal::SlotAllocator& tls)
+    : runtime_(runtime), dispatcher_(dispatcher), random_(random), local_info_(local_info),
+      tls_(tls) {}
+
+std::vector<Router::RouteConfigProviderSharedPtr> HttpRouteManagerImpl::routeConfigProviders() {
+  std::vector<Router::RouteConfigProviderSharedPtr> ret;
+  ret.reserve(route_config_providers_.size());
+  for (const auto& element : route_config_providers_) {
+    ret.push_back(element.second.lock());
```
Comment (htuch) [inline source/common/router/rds_impl.cc]: Do you want to include the failed locks here?

File: include/envoy/router/http_route_manager.h (hunk 10 above)
Comment (junr03) [inline include/envoy/router/http_route_manager.h]: this function will be used in a subsequent PR by a new admin handler. The handler will iterate
over all the configure RouteConfigProviders, and pretty print all the configured routes.

File: source/common/router/rds_impl.cc (hunk 12 above)
Comment (junr03) [inline source/common/router/rds_impl.cc]: That is what I was discussing with @mattklein123 in my initial comment above. I wanted to understand if the threading model allowed for failed locks or not. I understand that the way to protect here is to lock, check the returned ptr is nullptr, and then push only if it isn't, just wanted to make sure it was necessary.

File: source/common/router/rds_impl.cc (hunk 11 above)
Comment (junr03) [inline source/common/router/rds_impl.cc]: 1. Agreed, will do.
2. Good point. I guess that the correct way to do this is to have the map key be the tuple of the route_config_name and cluster.

File: source/common/router/rds_impl.cc (hunk 12 above)
Comment (mattklein123) [inline source/common/router/rds_impl.cc]: threading has nothing to do with it here. All of this is single threaded. I haven't looked at the code yet but assuming it's correct lock() should never return nullptr.

Comment (junr03) [inline source/common/router/rds_impl.cc]: Right, that's what I thought.

Comment (htuch) [inline source/common/router/rds_impl.cc]: Well, yeah, threading isn't relevant, `lock` is a misnomer. It's a way to turn weak into a real shared pointer. Assuming this is weak for a reason, it seems it should be possible for the `lock` to fail... or did I miss something here?

Comment (junr03) [inline source/common/router/rds_impl.cc]: I think I wasn't clear with my comment. I understand what lock does, and that it has nothing to do with threading. I was originally concerned that in the envoy threading model it would be possible to get to a state where the RouteConfigObject were destroyed but the weak ptr was still present in the Manager. I now understand that is not possible.

@htuch The pointer is weak because we want don't want the manager to own the RouteConfigProvider, we only want it to be able to dole out share ptrs to the HttpConnectionManagers that create RouteConfigProviders.

File: source/common/router/rds_impl.cc (hunk 11 above)
Comment (htuch) [inline source/common/router/rds_impl.cc]: Sure, or some other distinguished identifier (e.g. `:` can't appear in the cluster name.

File: source/common/router/rds_impl.cc [hunk 13]
```diff
@@ -119,5 +125,45 @@ void RdsRouteConfigProviderImpl::registerInitTarget(Init::Manager& init_manager)
...
+
+    new_provider->registerInitTarget(init_manager);
+
+    route_config_providers_.insert(
+        {map_identifier, std::weak_ptr<RdsRouteConfigProviderImpl>(new_provider)});
+
+    return std::move(new_provider);
+  }
+
+  return it->second.lock();
```
Comment (htuch) [inline source/common/router/rds_impl.cc]: Can this `lock()` fail? Based on other comment threads and the comment above about only cleaning up in `RouteConfigProvider` destructors, I'm thinking no. If that's the case, please add an ASSERT/comment to document. 

If it can fail here, shouldn't we have similar handling to the "not found" case above and create a new provider?

File: source/common/router/rds_impl.cc [hunk 14]
```diff
@@ -119,5 +125,45 @@ void RdsRouteConfigProviderImpl::registerInitTarget(Init::Manager& init_manager)
...
+    const LocalInfo::LocalInfo& local_info, ThreadLocal::SlotAllocator& tls)
+    : runtime_(runtime), dispatcher_(dispatcher), random_(random), local_info_(local_info),
+      tls_(tls) {}
+
+std::vector<Router::RouteConfigProviderSharedPtr>
+RouteConfigProviderManagerImpl::routeConfigProviders() {
+  std::vector<Router::RouteConfigProviderSharedPtr> ret;
+  ret.reserve(route_config_providers_.size());
+  for (const auto& element : route_config_providers_) {
+    ret.push_back(element.second.lock());
```
Comment (htuch) [inline source/common/router/rds_impl.cc]: OK, I think I get what's going on now. We're using a weak pointer simply as a means to allow handing out shared pointers later when we lock, otherwise we could be using plain `*` pointers. Seems legit. As below, I'd add an `ASSERT` here and comment to explain why this can never fail.

File: test/common/router/rds_impl_test.cc [hunk 15]
```diff
@@ -276,5 +278,80 @@ TEST_F(RdsImplTest, FailureArray) {
...
+
+  // Get a RouteConfigProvider. This one should create an entry in the RouteConfigProviderManager.
+  RouteConfigProviderSharedPtr provider = route_config_provider_manager_.getRouteConfigProvider(
+      *config, cm_, store_, "foo_prefix", init_manager_);
+  // Because this get has the same cluster and route_config_name, the provider returned is just a
+  // shared_ptr to the same provider as the one above.
+  RouteConfigProviderSharedPtr provider2 = route_config_provider_manager_.getRouteConfigProvider(
+      *config, cm_, store_, "foo_prefix", init_manager_);
+  // So this means that both shared_ptrs should be the same.
+  EXPECT_EQ(provider, provider2);
```
Comment (htuch) [inline test/common/router/rds_impl_test.cc]: Have you considered using the `shared_ptr` `use_count()` method to validate the behaviors around reference counting?

**Comment by mattklein123** [comment]
Please also merge master which will fix ASAN issue. I will take a look after you have addressed @htuch comments. Thanks.

**Comment by junr03** [comment]
@mattklein123 updated from comments. Merged master.

File: source/common/router/rds_impl.cc (hunk 11 above)
Comment (junr03) [inline source/common/router/rds_impl.cc]: @htuch concatenated with `:`. If we go this route then the `route_config_name` field of the RDS config also cannot contain `:`. I have documented that. But if I understand our breaking change policy I cannot enforce that until the 1.4.0 release. That means that when we release 1.4.0 I would make a PR enforcing the rule?

File: docs/configuration/http_conn_man/rds.rst [hunk 16]
```diff
@@ -28,7 +28,7 @@ route_config_name
   *(required, string)* The name of the route configuration. This name will be passed to the
   :ref:`RDS HTTP API <config_http_conn_man_rds_api>`. This allows an Envoy configuration with
   multiple HTTP listeners (and associated HTTP connection manager filters) to use different route
-  configurations.
+  configurations. The route_config_name must **not** contain ``:``.
```
Comment (htuch) [inline docs/configuration/http_conn_man/rds.rst]: Yeah, on second thoughts, maybe best to avoid baking in these dependencies. I know that we've been discussing trying to relax the `:` constraint on clusters at some point (but not now), so if there is a cleaner way, e.g. using tuples and hashing on that, then maybe do that.

**Comment by mattklein123** [comment]
Meta comment. I think this should probably use the singleton manager that I am building in https://github.com/lyft/envoy/pull/1410. This should merge today. This will avoid a lot of boilerplate addition for something that not everyone needs (for people that don't use HTTP). Even for admin endpoint, the singleton can register a handler into admin. Not a huge deal if we are eager for this to merge, but I might at least put in a TODO to this effect. I will review the rest of this change as is later today.

File: docs/configuration/http_conn_man/rds.rst (hunk 16 above)
Comment (mattklein123) [inline docs/configuration/http_conn_man/rds.rst]: Agreed. Don't think there is any explicit reason to block ':' ?

File: source/common/router/rds_impl.cc [hunk 17]
```diff
@@ -119,5 +125,56 @@ void RdsRouteConfigProviderImpl::registerInitTarget(Init::Manager& init_manager)
...
+RouteConfigProviderManagerImpl::routeConfigProviders() {
+  std::vector<Router::RouteConfigProviderSharedPtr> ret;
+  ret.reserve(route_config_providers_.size());
+  for (const auto& element : route_config_providers_) {
+    // Because the RouteConfigProviderManager's weak_ptrs only get cleaned up
+    // in the RdsRouteConfigProviderImpl destructor, and the single threaded nature
+    // of this code, locking the weak_ptr will not fail.
+    Router::RouteConfigProviderSharedPtr provider = element.second.lock();
+    ASSERT(provider)
+    ret.push_back(std::move(provider));
```
Comment (mattklein123) [inline source/common/router/rds_impl.cc]: del std::move here as this is a shared_ptr

File: source/common/router/rds_impl.cc [hunk 18]
```diff
@@ -119,5 +125,56 @@ void RdsRouteConfigProviderImpl::registerInitTarget(Init::Manager& init_manager)
...
+  return ret;
+};
+
+Router::RouteConfigProviderSharedPtr RouteConfigProviderManagerImpl::getRouteConfigProvider(
+    const Json::Object& config, Upstream::ClusterManager& cm, Stats::Scope& scope,
+    const std::string& stat_prefix, Init::Manager& init_manager) {
+
+  // RdsRouteConfigProviders are unique based on their <route_config_name>_<cluster>.
+  const std::string manager_identifier =
+      config.getString("route_config_name") + ":" + config.getString("cluster");
```
Comment (mattklein123) [inline source/common/router/rds_impl.cc]: The key is opaque. If you just remove ':' and then you can allow ':' in the name (you probably could allow ':' in the name anyway and it would work).

File: source/common/router/rds_impl.cc [hunk 19]
```diff
@@ -119,5 +125,56 @@ void RdsRouteConfigProviderImpl::registerInitTarget(Init::Manager& init_manager)
...
+  auto it = route_config_providers_.find(manager_identifier);
+  if (it == route_config_providers_.end()) {
+    std::shared_ptr<RdsRouteConfigProviderImpl> new_provider{
+        new RdsRouteConfigProviderImpl(config, manager_identifier, runtime_, cm, dispatcher_,
+                                       random_, local_info_, scope, stat_prefix, tls_, *this)};
+
+    new_provider->registerInitTarget(init_manager);
+
+    route_config_providers_.insert(
+        {manager_identifier, std::weak_ptr<RdsRouteConfigProviderImpl>(new_provider)});
```
Comment (mattklein123) [inline source/common/router/rds_impl.cc]: nit: I don't think`std::weak_ptr<RdsRouteConfigProviderImpl>` is needed here as it should auto convert.

File: source/common/router/rds_impl.cc [hunk 20]
```diff
@@ -119,5 +125,56 @@ void RdsRouteConfigProviderImpl::registerInitTarget(Init::Manager& init_manager)
...
+
+    return std::move(new_provider);
+  }
+
+  // Because the RouteConfigProviderManager's weak_ptrs only get cleaned up
+  // in the RdsRouteConfigProviderImpl destructor, and the single threaded nature
+  // of this code, locking the weak_ptr will not fail.
+  Router::RouteConfigProviderSharedPtr new_provider = it->second.lock();
+  ASSERT(new_provider);
+  return std::move(new_provider);
```
Comment (mattklein123) [inline source/common/router/rds_impl.cc]: del std::move

Comment (junr03) [inline source/common/router/rds_impl.cc]: @mattklein123 for my information. Here and above, doesn't moving the pointer instead of not save the atomic operations of copying the pointer on return (here) and insertion into the vector?

File: docs/configuration/http_conn_man/rds.rst (hunk 16 above)
Comment (junr03) [inline docs/configuration/http_conn_man/rds.rst]: Thinking about this some more, I was being near sighted, and thought that it was possible to fall into the same collision situation that I was trying to escape. But that is not the case if one of the strings has the blocked character. 

@htuch how serious are we thinking about relaxing the `:` constraint? I could leave a TODO to move the to pair, or struct key when that happens, but for now I'll leave it this way to avoid having to deal with hashing the key.

Comment (mattklein123) [inline docs/configuration/http_conn_man/rds.rst]: I don't really understand how ':' matters here at all. The key is internal. If you just add config name + cluster, what difference does it make what characters it contains?

Comment (junr03) [inline docs/configuration/http_conn_man/rds.rst]: Because if for example someone specifies route_config_name `foo` and cluster `bar` and someone else does route_config_name `foob` and cluster `ar` the concatenation would make the two providers collide when it shouldn't.

File: source/common/router/rds_impl.cc (hunk 20 above)
Comment (junr03) [inline source/common/router/rds_impl.cc]: ok, got it.

File: docs/configuration/http_conn_man/rds.rst (hunk 16 above)
Comment (junr03) [inline docs/configuration/http_conn_man/rds.rst]: You probably see that. So I am guessing I am the one that doesn't see why this wouldn't matter? I guess it is pretty contrived that such a collision would happen, but I think protecting against it is better?

Comment (mattklein123) [inline docs/configuration/http_conn_man/rds.rst]: Then just do something like "hash_separator_you_need_to_be_really_insane_to_use_this" or whatever. (I haven't thought through it enough but seems like something like this should work fine.

The other option is to just do a custom hasher/comparer which operates on a struct or tuple of config/cluster.

**Comment by junr03** [comment]
@mattklein123 for sake of finishing this PR, I am leaning into merging in the current state and doing a subsequent PR to use #1410.

File: source/common/router/rds_impl.cc [hunk 21]
```diff
@@ -119,5 +125,55 @@ void RdsRouteConfigProviderImpl::registerInitTarget(Init::Manager& init_manager)
...
+    const std::string& stat_prefix, Init::Manager& init_manager) {
+
+  // RdsRouteConfigProviders are unique based on their <route_config_name>MAP_CONCATENATOR<cluster>.
+  const std::string manager_identifier =
+      config.getString("route_config_name") + MAP_CONCATENATOR + config.getString("cluster");
+
+  auto it = route_config_providers_.find(manager_identifier);
+  if (it == route_config_providers_.end()) {
+    std::shared_ptr<RdsRouteConfigProviderImpl> new_provider{
+        new RdsRouteConfigProviderImpl(config, manager_identifier, runtime_, cm, dispatcher_,
```
Comment (htuch) [inline source/common/router/rds_impl.cc]: Nit: `make_shared`

File: test/common/router/rds_impl_test.cc [hunk 22]
```diff
@@ -276,5 +278,85 @@ TEST_F(RdsImplTest, FailureArray) {
   EXPECT_EQ(1UL, store_.counter("foo.rds.update_failure").value());
 }
 
+class RouteConfigProviderManagerImplTest : public testing::Test {
+public:
+  RouteConfigProviderManagerImplTest() {}
```
Comment (mattklein123) [inline test/common/router/rds_impl_test.cc]: nit: del empty constructor/destructor

File: test/common/router/rds_impl_test.cc [hunk 23]
```diff
@@ -276,5 +278,85 @@ TEST_F(RdsImplTest, FailureArray) {
...
+  EXPECT_EQ(2UL, configured_providers.size());
+  EXPECT_EQ(3UL, provider.use_count());
+  EXPECT_EQ(2UL, provider3.use_count());
+
+  provider.reset();
+  provider2.reset();
+  configured_providers.clear();
+
+  // All shared_ptrs to the provider pointed at by provider1, and provider2 have been deleted, so
+  // now we should only have the provider porinted at by provider3.
```
Comment (mattklein123) [inline test/common/router/rds_impl_test.cc]: typo "porinted"

**Comment by junr03** [comment]
@mattklein123 updated.

**Comment by mattklein123** [comment]
@junr03 still need to fix @htuch nit.
